dds/idl/marshal_generator.cpp

```diff
@@ -1915,26 +1915,18 @@ bool marshal_generator::gen_struct(AST_Structure* node,
       }
     }
     if (may_be_parameter_list) {
-      if (repr.xcdr2) {
         /**
          * We don't have a PID marking the end in XCDR2 parameter lists, but we
          * have the size, so we need to stop after we hit this offset.
-         *
-         * TODO(iguessthislldo): Replace this with a cleaner mechanism where
-         * the Serializer keeps track of the stream offset.
-         * https://github.com/objectcomputing/OpenDDS/pull/1722#discussion_r447056830
          */
-        be_global->impl_ <<
-          "  const char* end_of_fields = strm.pos_rd() + total_size;\n";
-      }
       be_global->impl_ <<
         "  unsigned member_id;\n"
         "  size_t field_size;\n"
         "  while (true) {\n";
 
       if (repr.xcdr2) {
         be_global->impl_ << "\n"
-          "    if ((!strm.pos_rd() || strm.pos_rd() >= end_of_fields)";
+          "    if (strm.pos() >= total_size + uint32_cdr_size";
```

Comment:
The struct doesn't have to be the topic type and may be nested, so may not start at `strm.pos() == 0`, so you still need `end_of_fields` or something like it. Also what is `uint32_cdr_size`?

Reply:
Thanks. `uint32_cdr_size` is for the size of DHEADER as `total_size` doesn't include the DHEADER's size. But that assumes `strm.pos()` starts at 0 when reading this struct. So I'll need to change this code.

Reply:
The `DHEADER` should already be included in the result of `serialized_size`.

Reply:
I think `total_size` is only the size of the payload data. `read_delimiter` writes the value of DHEADER to `total_size` and DHEADER doesn't include the size of DHEADER itself.

Reply:
You're right, I got mixed up. Also `start_pos` is before `read_delimter`, where `end_of_fields` was afterwards. If you move `start_pos` to after `read_delimter`, I don't think you have to add the DHEADER size anymore.

Reply:
Right. I've changed this in the latest commit.